### DISCUSSION TITLE
❓ Question: Correct way to use action

### DISCUSSION THREAD

**Discussion by dcheglakov** [melt-ui/melt-ui#106] [Q&A] [answered]
Starting from version 0.13.0 we need to use `use:action` for some elements.
What is the correct way to do this for reusable components?

For example, I have a `Close button` component:
```js
<script lang="ts">
    import LucideX from '~icons/lucide/x';
</script>

<button
    {...$$props}
    aria-label="Close"
    class=" rounded-full p-2 text-slate-600 hover:bg-slate-200 hover:text-slate-800"
>
    <LucideX />
</button>
```

To make my button work, I did the following:
```js
<script lang="ts">
    import LucideX from '~icons/lucide/x';
    export let action: (node: HTMLButtonElement) => { destroy: () => void } = () => ({
		destroy: () => {}
	});
</script>

<button
    {...$$props}
    aria-label="Close"
    use:action
    class=" rounded-full p-2 text-slate-600 hover:bg-slate-200 hover:text-slate-800"
>
    <LucideX />
</button>
```

And I use it like this:
```js
<CloseButton {...close} action={close.action} />
```

Is it right, or is there a better way?
Thanks.

**Comment by TGlide** [comment, marked as answer]
Hi! That is a possible implementation of it, yes. You may also want to accept an action array, using a helper: https://github.com/radix-svelte/radix-svelte/blob/develop/src/lib/internal/helpers/useActions.ts

Also, the action type can potentially be simpler, by importing:
`import { type Action } from 'svelte/action';`

Closing it due to it not being a bug, but feel free to continue discussing it. Alternatively, open a discussion in the repo: https://github.com/melt-ui/melt-ui/discussions

**Reply by TGlide** [reply]
EDIT: Converted to discussion!

**Reply by dcheglakov** [reply]
Thanks `import type { Action } from 'svelte/action';` indeed simpler.
I don't quite understand `useActions` yet, it's a bit difficult for me, I need to figure it out.

I also wanted to ask if it's normal for these actions to be included in HTML?

Here is resulted `HTML` in `dev`:
```html
<button type="button" action="(node) => {
      const unsub = addEventListener(node, &quot;click&quot;, () => {
        open.set(false);
      });
      return {
        destroy: unsub
      };
    }" class="rounded-full p-2 text-slate-600 hover:bg-slate-200 hover:text-slate-800"><svg viewBox="0 0 24 24" width="1.2em" height="1.2em"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M18 6L6 18M6 6l12 12"></path></svg><!--<X>--></button>
```

and in `prod`:
```html
<button type="button" action="f=>({destroy:M(f,&quot;click&quot;,()=>{o.set(!1)})})" class="rounded-full p-2 text-slate-600 hover:bg-slate-200 hover:text-slate-800"><svg viewBox="0 0 24 24" width="1.2em" height="1.2em"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M18 6L6 18M6 6l12 12"></path></svg></button>
```

**Reply by pawelblaszczyk5** [reply]
$$props include the declared props as well. Hence, you're spreading the action onto button as well. You could use $$restProps here, the difference is that this one doesn't include props that you declared with `export let`
